scylla/src/transport/session.rs

```diff
@@ -1821,77 +1821,33 @@ impl Session {
         last_error.map(Result::Err)
     }
 
-    pub async fn await_schema_agreement(&self) -> Result<(), QueryError> {
-        while !self.check_schema_agreement().await? {
-            tokio::time::sleep(self.schema_agreement_interval).await
+    async fn await_schema_agreement_indefinitely(&self) -> Result<Uuid, QueryError> {
+        loop {
+            tokio::time::sleep(self.schema_agreement_interval).await;
+            if let Some(agreed_version) = self.check_schema_agreement().await? {
+                return Ok(agreed_version);
+            }
         }
-        Ok(())
     }
 
-    pub async fn await_timed_schema_agreement(
+    pub async fn await_schema_agreement(
         &self,
         timeout_duration: Duration,
-    ) -> Result<bool, QueryError> {
-        timeout(timeout_duration, self.await_schema_agreement())
+    ) -> Result<Option<Uuid>, QueryError> {
```

Comment:
```suggestion
    ) -> Result<Uuid, QueryError> {
```

maybe the timeout should be represented as a timeout `QueryError`?

Reply:
Good idea!

Reply:
The only problem I see here is that we lose a way to conveniently distinguish cases of timeout (not yet in agreement) from cases of server-side or connection-related errors.

Reply:
I've implemented it anyway.